File: src/jray/ray/scenes/RandomForest.java
package jray.ray.scenes;

import java.io.IOException;
import java.util.List;
import java.util.Vector;

import jray.common.Cone;
import jray.common.Object3D;
import jray.common.Sky;
import jray.common.Sphere;
import jray.common.Vect3;
import jray.math.intersections.IntColors;
import jray.ray.Camera;
import jray.ray.Scene;
import jray.struct.Octree;


public class RandomForest extends Scene {

    private List<Object3D> objects = new Vector<Object3D>();
    private Camera c;
    private Octree tree;

    public RandomForest() throws IOException{
    	double dist=0.2;
    	double h = 0.3;
    	double camheight = 10;
    	
    	for(double x = 1;x<camheight*3;x+=h)
    		for(double z = -camheight;z<camheight;z+=h)
    			addTree(x+Math.random()*dist-dist/2,-2,z+Math.random()*dist-dist/2);
        
        c = new Camera(new Vect3(-camheight, camheight*2, 0), new Vect3(-camheight+1, camheight*2-1, 0), new Vect3(0.707106781186547, 0.707106781186547, 0), 1, 1, 1);
        objects.add(c);
        
        tree = Octree.buildTree(new Vect3(camheight*1.5,-3.1,0),objects);
    }

    private void addTree(double x, double y, double z){
    	double dist= 0.1;
    	double t0 = Math.random()*dist-dist/2;
    	double t1 = Math.random()*dist-dist/2;
    	double ld = 0.2;
    	
    	int leafColor = IntColors.A+(((int)(IntColors.G*(0.5 + 0.5*Math.random())))&IntColors.G);
    	
    	if(Math.random()<0.5){
    		objects.add(new Cone(new Vect3(x,y+2,z),new Vect3(t0,-1.6,t1),10,leafColor));
    		objects.add(new Cone(new Vect3(x,y+2,z),new Vect3(t0*2,-2,t1*2),1,0xFF8B4513));
    	}else{
    		objects.add(new Sphere(new Vect3(x,y+1.5-ld/2,z),ld,leafColor));
    		objects.add(new Cone(new Vect3(x,y+1.5,z),new Vect3(t0*2,-1.5,t1*2),1,0xFF8B4513));
    	}
    }
    
    @Override
    public List<Object3D> getObjects() {
        return objects;
    }

    @Override
    public Camera getCamera() {
        return c;
    }

    @Override
    public Sky getSky() {
        return null;
    }
    
    @Override
    public Octree getSceneTree(){
    	return tree;
    }
}

File: JRayXLib/JRayXLib/Ray/Scenes/RandomForest.cs
using System;
using System.Collections.Generic;
using JRayXLib.Colors;
using JRayXLib.Shapes;
using JRayXLib.Struct;

namespace JRayXLib.Ray.Scenes
{
    public class RandomForest : Scene
    {
        private static readonly Random Rd = new Random();
        private readonly Octree _tree;

        public RandomForest()
        {
            const double dist = 0.2;
            const double h = 0.3;
            const double camheight = 10;

            var objects = new List<I3DObject>();

            for (double x = 1; x < camheight * 3; x += h)
                for (double z = -camheight; z < camheight; z += h)
                    AddTree(x + Rd.NextDouble() * dist - dist / 2, -2, z + Rd.NextDouble() * dist - dist / 2, objects);

            Camera = new Camera(
                new Vect3 { X = -camheight, Y = camheight * 2 },
                new Vect3 { X = -camheight + 1, Y = camheight * 2 - 1 },
                new Vect3 { X = 0.707106781186547, Y = 0.707106781186547 }, 1, 1, 1);
            objects.Add(Camera);
            Objects = objects.ToArray();

            _tree = Octree.BuildTree(new Vect3 { X = camheight * 1.5, Y = -3.1 }, Objects);
        }

        private void AddTree(double x, double y, double z, IList<I3DObject> objects)
        {
            const double dist = 0.1;
            double t0 = Rd.NextDouble() * dist - dist / 2;
            double t1 = Rd.NextDouble() * dist - dist / 2;
            const double ld = 0.2;

            var leafColor = new Color
                {
                    A = byte.MaxValue,
                    R = byte.MaxValue / 3,
                    G = (byte)(byte.MaxValue * ((1 + Rd.NextDouble()) / 2)),
                    B = byte.MaxValue / 3
                };
            var @base = new Color
                {
                    A = byte.MaxValue,
                    R = 0x8B,
                    G = 0x45,
                    B = 0x13
                };

            if (Rd.NextDouble() < 0.5)
            {
                objects.Add(new Cone(new Vect3 { X = x, Y = y + 2, Z = z }, new Vect3 { X = t0, Y = -1.6, Z = t1 }, 10, leafColor));
                objects.Add(new Cone(new Vect3 { X = x, Y = y + 2, Z = z }, new Vect3 { X = t0 * 2, Y = -2, Z = t1 * 2 }, 1, @base));
            }
            else
            {
                objects.Add(new Sphere(new Vect3 { X = x, Y = y + 1.5 - ld / 2, Z = z }, ld, leafColor));
                objects.Add(new Cone(new Vect3 { X = x, Y = y + 1.5, Z = z }, new Vect3 { X = t0 * 2, Y = -1.5, Z = t1 * 2 }, 1, @base));
            }
        }

        public override Sky GetSky()
        {
            return null;
        }

        public override Octree GetSceneTree()
        {
            return _tree;
        }
    }
}